### ISSUE TITLE
Integration tests can fail waiting for the stack to become ready

### ISSUE THREAD

**Issue by cmacknz** [elastic/elastic-agent#3167]
* **Link:** https://buildkite.com/elastic/elastic-agent/builds/2203#0189b271-9d72-489e-86b8-1b2ccc225077
* **Branch:** main
* **Artifact Link:** https://buildkite.com/elastic/elastic-agent/builds/2203#0189b271-9d72-489e-86b8-1b2ccc225077

### Stack Trace

```
2023-08-01 15:07:04 EDT	>>> (linux-arm64-ubuntu-2204) Failed for instance linux-arm64-ubuntu-2204 (@ 35.226.36.231): ogc-linux-arm64-ubuntu-2204-6866 unable to continue because stack never became ready: failed to check for cloud 8.10.0-SNAPSHOT to be ready: error calling deployment retrieval API: Get "https://console.qa.cld.elstc.co/api/v1/deployments/cc8822af466e49b5879d25f7c2b7bf1d": context deadline exceeded
2023-08-01 15:07:04 EDT	
2023-08-01 15:07:18 EDT	>>> (linux-arm64-ubuntu-2204-teststandaloneupgrade) Copying agent build elastic-agent-8.10.0-SNAPSHOT-linux-arm64.tar.gz
2023-08-01 15:07:20 EDT	>>> (linux-arm64-ubuntu-2204-teststandaloneupgrade) Failed to copy files instance: failed to create agent/build/distributions directory: context canceled (stdout: , stderr: )
2023-08-01 15:07:20 EDT	>>> (linux-arm64-ubuntu-2204-teststandaloneupgrade) Failed for instance linux-arm64-ubuntu-2204-teststandaloneupgrade (@ 35.188.124.245): failed to copy files to instance ogc-linux-arm64-ubuntu-2204-teststandaloneupgrade-892d: failed to create agent/build/distributions directory: context canceled (stdout: , stderr: )
2023-08-01 15:07:20 EDT	
2023-08-01 15:08:27 EDT	>>> (linux-amd64-ubuntu-2204-teststandaloneupgrade) Copying agent build elastic-agent-8.10.0-SNAPSHOT-linux-x86_64.tar.gz
2023-08-01 15:08:29 EDT	>>> (linux-amd64-ubuntu-2204-teststandaloneupgrade) Failed to copy files instance: failed to create agent/build/distributions directory: context canceled (stdout: , stderr: )
2023-08-01 15:08:29 EDT	>>> (linux-amd64-ubuntu-2204-teststandaloneupgrade) Failed for instance linux-amd64-ubuntu-2204-teststandaloneupgrade (@ 34.31.157.132): failed to copy files to instance ogc-linux-amd64-ubuntu-2204-teststandaloneupgrade-63af: failed to create agent/build/distributions directory: context canceled (stdout: , stderr: )
2023-08-01 15:08:29 EDT	
2023-08-01 15:08:43 EDT	>>> (linux-amd64-ubuntu-2204) Copying agent build elastic-agent-8.10.0-SNAPSHOT-linux-x86_64.tar.gz
2023-08-01 15:08:46 EDT	>>> (linux-amd64-ubuntu-2204) Failed to copy files instance: failed to create agent/build/distributions directory: context canceled (stdout: , stderr: )
2023-08-01 15:08:46 EDT	>>> (linux-amd64-ubuntu-2204) Failed for instance linux-amd64-ubuntu-2204 (@ 34.173.223.7): failed to copy files to instance ogc-linux-amd64-ubuntu-2204-06f8: failed to create agent/build/distributions directory: context canceled (stdout: , stderr: )
2023-08-01 15:08:46 EDT	
2023-08-01 15:08:46 EDT	Error: error running test: ogc-linux-arm64-ubuntu-2204-6866 unable to continue because stack never became ready: failed to check for cloud 8.10.0-SNAPSHOT to be ready: error calling deployment retrieval API: Get "https://console.qa.cld.elstc.co/api/v1/deployments/cc8822af466e49b5879d25f7c2b7bf1d": context deadline exceeded
```

Possibly we just need to increase the timeout, it's not clear if the deployment was actually created or not or if it just timed out just from looking at the logs.

**Comment by blakerouse** [comment]
Yeah I have been seeing the stack fail to become ready more recently even on my local usage of the integration testing framework. For me it was because the stack really didn't become ready, even on a new deployment. It was marked Unhealthy.

**Comment by cmacknz** [comment]
Interesting, is there any way to get the stack logs when this happens?

This could be an artifact of the QA environment not being as stable as production Elastic Cloud as well, perhaps we just need to try creating the stack instead? Hard to say without knowing exactly what went wrong, or how often this happens.

**Comment by ycombinator** [comment]
I looked through https://www.elastic.co/guide/en/cloud/current/ec-restful-api.html and I don't see any API that would let us get logs for a deployment.  However, I do know that deployment logs get shipped to an internal ES cluster (called a regional logging cluster or RLC).  Perhaps if the test runner could print out the deployment ID as well as the cluster IDs of the deployment components - ES, Kibana, Fleet Server, that could at least help us find these logs on the appropriate RLC.

**Comment by cmacknz** [comment]
So the tests do log the deployment ID as it is part of the failing API request path:

```
Error: error running test: ogc-linux-arm64-ubuntu-2204-d5b1 unable to continue because stack never became ready: failed to check for cloud 8.9.1-SNAPSHOT to be ready: error calling deployment retrieval API: Get "https://console.qa.cld.elstc.co/api/v1/deployments/288a4ec3e34a4658a886566f4d576f66": context deadline exceeded
--
  | 2023-08-03 08:34:42 EDT | + TESTS_EXIT_STATUS=1
  | 2023-08-03 08:34:42 EDT | + set -e
```

If I go to the QA cloud admin console this is what I see:

<img width="813" alt="Screen Shot 2023-08-03 at 11 22 32 AM" src="https://github.com/elastic/elastic-agent/assets/3466215/5249c1c4-c87e-4d54-b45d-5503fcc33809">

So it did eventually become healthy, but with a warning:

<img width="1474" alt="Screen Shot 2023-08-03 at 11 22 48 AM" src="https://github.com/elastic/elastic-agent/assets/3466215/8295fb52-86e6-433f-bb19-f2ffb6caf66b">

Is this "Healthy, with warnings" status something we need to explicitly handle as a response from the deployment API? Seems like the warning here can be ignored.

Alternatively if we get this it might be easier to just create another deployment, but I do see at least one other ingest-ci deployment that failed due to a lack of capacity so that could make things worse.

**Comment by ycombinator** [comment]
> Is this "Healthy, with warnings" status something we need to explicitly handle as a response from the deployment API? Seems like the warning here can be ignored.

Yes, this warning can be ignored.  All our deployments are created [with a single zone](https://github.com/elastic/elastic-agent/blob/a7bd2f07590370721794e877d38b6fc42b89f855/pkg/testing/ess/deployment.go#L345) for the ES cluster.  This is a deliberate choice so the deployment spins up as fast as possible.  We sacrifice redundancy but I figured that was a reasonable tradeoff given that we're just using these deployments for testing purposes and they're intended to be short-lived.

Once the deployment ID is known, it's possible to visit the QA Admin Console, visit the deployment's page there, e.g. https://admin.qa.cld.elstc.co/deployments/288a4ec3e34a4658a886566f4d576f66, then visit the ES and Kibana pages under that deployment (links in the left nav), click on External Links on each of those pages, and follow the link to the cluster logs for each.  Screenshots below.


<img width="1036" alt="Screenshot 2023-08-03 at 08 40 37" src="https://github.com/elastic/elastic-agent/assets/51061/81205b24-3281-456a-b416-317382bd162c">

<img width="1134" alt="Screenshot 2023-08-03 at 08 40 27" src="https://github.com/elastic/elastic-agent/assets/51061/faabd2fb-79a0-4295-b2fe-354a5bb85983">
